### PR TITLE
[GLUTEN-6571][VL] Add platform and arch subdirectory for base lib package

### LINKED ISSUE #6571: [VL] Add an arch subdirectory for native libs package
### Description

There are machines with different CPU architectures in our internal hybrid cloud cluster, so I want to add an arch subdirectory for native libs package.

### PR DESCRIPTION
**Pull request by wForget**
## What changes were proposed in this pull request?

Add platform and arch subdirectory for base lib package.

closes #6571

## How was this patch tested?

Manual package and test.

Check gluten bundle jar:

```
jar -tf gluten-velox-bundle-spark3.5_2.12-centos_7_x86_64-1.3.0-SNAPSHOT.jar | grep -E "*.so$"

# output
org/apache/gluten/linux/amd64/libgluten.so
org/apache/gluten/linux/amd64/libvelox.so
......
```

Successfully load native libs:

```
24/08/21 12:44:53 INFO JniLibLoader: Successfully loaded library org/apache/gluten/linux/amd64/libgluten.so
24/08/21 12:44:54 INFO JniLibLoader: Successfully loaded library org/apache/gluten/linux/amd64/libvelox.so
```

Successfully execute tpcds q3.

### CODE REVIEW

**Review by zhztheplayer — COMMENTED** [review]
Looking good though does this work with 3rd libs?

https://github.com/apache/incubator-gluten/blob/11a236dd86f2f25777f2df9db9f1fef4f71e97f7/dev/build-thirdparty.sh#L85C1-L85C62

It's likely the libs are already packaged into specific jars. So I assume we don't have to anything for that?

cc @PHILO-HE

**Comment by wForget** [comment]
With vpkg packaging, `3rd libs` seem to be no longer needed? Do we still need to handle `3rd libs` in this way?

**Comment by zhztheplayer** [comment]
> > Looking good though does this work with 3rd libs?
> 
> With vpkg packaging, `3rd libs` seem to be no longer needed? Do we still need to handle `3rd libs` in this way?

I personally never used that feature. @PHILO-HE Do you know some? It would sound great if we can remove, as for long term we will completely drop dynamic build support.

File: gluten-core/src/main/java/org/apache/gluten/vectorized/JniLibLoader.java [hunk 1]
```diff
@@ -159,10 +160,16 @@ public static void unloadFromPath(String libPath) {
   }
 
   public void mapAndLoad(String unmappedLibName, boolean requireUnload) {
+    mapAndLoad(Optional.empty(), unmappedLibName, requireUnload);
+  }
+
+  public void mapAndLoad(
+      Optional<String> packagePath, String unmappedLibName, boolean requireUnload) {
```
Comment (zhztheplayer) [inline gluten-core/src/main/java/org/apache/gluten/vectorized/JniLibLoader.java:167]: Would suggest keeping one single `mapAndLoad` then doing the concatenation in caller code.

Probably just rename the arg in previous `mapAndLoad` method from `unmappedLibName` to `unmappedLibPath`?

**Review by zhztheplayer — COMMENTED** [review]
👍

File: gluten-core/src/main/java/org/apache/gluten/vectorized/JniLibLoader.java (hunk 1 above)
Comment (wForget) [reply gluten-core/src/main/java/org/apache/gluten/vectorized/JniLibLoader.java:167]: Makes sense, I will do that

Comment (wForget) [reply gluten-core/src/main/java/org/apache/gluten/vectorized/JniLibLoader.java:167]: Can the `System.mapLibraryName(unmappedLibName)` inside `mapAndLoad` method be moved to the calling method?

Comment (zhztheplayer) [reply gluten-core/src/main/java/org/apache/gluten/vectorized/JniLibLoader.java:167]: > Can the `System.mapLibraryName(unmappedLibName)` inside `mapAndLoad` method be moved to the calling method?

Sounds good to me.

Comment (wForget) [reply gluten-core/src/main/java/org/apache/gluten/vectorized/JniLibLoader.java]: Changed, thank you

File: gluten-core/src/main/java/org/apache/gluten/vectorized/JniLibLoader.java [hunk 2]
```diff
@@ -159,17 +158,10 @@ public static void unloadFromPath(String libPath) {
     }
   }
 
-  public void mapAndLoad(String unmappedLibName, boolean requireUnload) {
-    mapAndLoad(Optional.empty(), unmappedLibName, requireUnload);
-  }
-
-  public void mapAndLoad(
-      Optional<String> packagePath, String unmappedLibName, boolean requireUnload) {
+  public void mapAndLoad(String unmappedLibPath, boolean requireUnload) {
```
Comment (zhztheplayer) [inline gluten-core/src/main/java/org/apache/gluten/vectorized/JniLibLoader.java:161]: Could remove the method.

File: gluten-core/src/main/java/org/apache/gluten/vectorized/JniLibLoader.java (hunk 1 above)
Comment (zhztheplayer) [reply gluten-core/src/main/java/org/apache/gluten/vectorized/JniLibLoader.java]: And can we do the same on `loadAndCreateLink` too? For consistency. Though I assume that one requires for some refactors in caller code.

**Comment by wForget** [comment]
@zhztheplayer @PHILO-HE  Could you please help to see if the [failed GA](https://opencicd.kyligence.com/job/gluten/job/gluten-ci/11653/) is a relevant failure? 

```
16:32:19  - test cache file command *** FAILED ***
16:32:19    org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 4.0 failed 1 times, most recent failure: Lost task 0.0 in stage 4.0 (TID 4) (gluten-gluten-ci-11653-llqgx-65q8j-rm2bs executor driver): org.apache.gluten.exception.GlutenException: org.apache.gluten.exception.GlutenException: org.apache.gluten.exception.GlutenException: Open file(hdfs://127.0.0.1:8020/tpch-data/lineitem/part-00000-d08071cb-0dfa-42dc-9198-83cb334ccda3-c000.snappy.parquet) failed. IOError: Code: 107. DB::ErrnoException: Cannot open file /tmp/gluten_hdfs_cache//4654/6d2/6d25666d180cc924b04f3234d772995f/0: , errno: 2, strerror: No such file or directory: Cache info: Buffer path: tpch-data/lineitem/part-00000-d08071cb-0dfa-42dc-9198-83cb334ccda3-c000.snappy.parquet, hash key: 6d25666d180cc924b04f3234d772995f, file_offset_of_buffer_end: 19164575, read_until_position: 19230111, internal buffer end: None, read_type: CACHED, last caller: 215fdc3e-15d9-4926-b9ad-9201b2f74587:57933, file segment info: File segment: [0, 19230110], key: 6d25666d180cc924b04f3234d772995f, state: DOWNLOADED, downloaded size: 19230111, reserved size: 19230111, downloader id: None, current write offset: 19230111, caller id: 215fdc3e-15d9-4926-b9ad-9201b2f74587:57933, kind: Regular, unbound: 0. (FILE_DOESNT_EXIST) (version 24.9.1.1): While executing SubstraitFileSource
16:32:19  0. /home/jenkins/agent/workspace/gluten/gluten-ci/gluten/cpp-ch/build/../ClickHouse/src/Common/Exception.cpp:111: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000bc09edb
16:32:19  1. ../ClickHouse/src/Common/Exception.h:110: DB::Exception::Exception(PreformattedMessage&&, int) @ 0x00000000064a3f2c
16:32:19  2. ../ClickHouse/src/Common/Exception.h:128: DB::Exception::Exception<String const&, String>(int, FormatStringHelperImpl<std::type_identity<String const&>::type, std::type_identity<String>::type>, String const&, String&&) @ 0x00000000064c036b
16:32:19  3. /home/jenkins/agent/workspace/gluten/gluten-ci/gluten/cpp-ch/build/../local-engine/Storages/SubstraitSource/ParquetFormatFile.cpp:146: local_engine::ParquetFormatFile::collectRequiredRowGroups(DB::ReadBuffer*, int&) const @ 0x000000000c30ee5f
16:32:19  4. /home/jenkins/agent/workspace/gluten/gluten-ci/gluten/cpp-ch/build/../local-engine/Storages/SubstraitSource/ParquetFormatFile.cpp:70: local_engine::ParquetFormatFile::createInputFormat(DB::Block const&) @ 0x000000000c30d3e6
16:32:19  5. /home/jenkins/agent/workspace/gluten/gluten-ci/gluten/cpp-ch/build/../local-engine/Storages/SubstraitSource/SubstraitFileSource.cpp:345: local_engine::NormalFileReader::NormalFileReader(std::shared_ptr<local_engine::FormatFile> const&, std::shared_ptr<DB::Context const> const&, DB::Block const&, DB::Block const&) @ 0x000000000c3059f2
16:32:19  6. ../ClickHouse/contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:714: local_engine::SubstraitFileSource::generate() @ 0x000000000c303049
16:32:19  7. /home/jenkins/agent/workspace/gluten/gluten-ci/gluten/cpp-ch/build/../ClickHouse/src/Processors/ISource.cpp:139: DB::ISource::tryGenerate() @ 0x000000001013ac5b
16:32:19  8. /home/jenkins/agent/workspace/gluten/gluten-ci/gluten/cpp-ch/build/../ClickHouse/src/Processors/ISource.cpp:108: DB::ISource::work() @ 0x000000001013a967
16:32:19  9. /home/jenkins/agent/workspace/gluten/gluten-ci/gluten/cpp-ch/build/../ClickHouse/src/Processors/Executors/ExecutionThreadContext.cpp:47: DB::ExecutionThreadContext::executeTask() @ 0x00000000101550c9
16:32:19  10. /home/jenkins/agent/workspace/gluten/gluten-ci/gluten/cpp-ch/build/../ClickHouse/src/Processors/Executors/PipelineExecutor.cpp:279: DB::PipelineExecutor::executeStepImpl(unsigned long, std::atomic<bool>*) @ 0x0000000010148570
16:32:19  11. /home/jenkins/agent/workspace/gluten/gluten-ci/gluten/cpp-ch/build/../ClickHouse/src/Processors/Executors/PipelineExecutor.cpp:153: DB::PipelineExecutor::executeStep(std::atomic<bool>*) @ 0x0000000010147f48
16:32:19  12. /home/jenkins/agent/workspace/gluten/gluten-ci/gluten/cpp-ch/build/../ClickHouse/src/Processors/Executors/PullingPipelineExecutor.cpp:54: DB::PullingPipelineExecutor::pull(DB::Chunk&) @ 0x000000001015ba77
16:32:19  13. /home/jenkins/agent/workspace/gluten/gluten-ci/gluten/cpp-ch/build/../ClickHouse/src/Processors/Executors/PullingPipelineExecutor.cpp:65: DB::PullingPipelineExecutor::pull(DB::Block&) @ 0x000000001015bc99
16:32:19  14. /home/jenkins/agent/workspace/gluten/gluten-ci/gluten/cpp-ch/build/../local-engine/Parser/SerializedPlanParser.cpp:1613: local_engine::LocalExecutor::hasNext() @ 0x000000000c030e45
16:32:19  15. /home/jenkins/agent/workspace/gluten/gluten-ci/gluten/cpp-ch/build/../local-engine/local_engine_jni.cpp:277: Java_org_apache_gluten_vectorized_BatchIterator_nativeHasNext @ 0x0000000006487e77
16:32:19  
16:32:19  0. /home/jenkins/agent/workspace/gluten/gluten-ci/gluten/cpp-ch/build/../ClickHouse/src/Common/Exception.cpp:111: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000bc09edb
16:32:19  1. ../ClickHouse/src/Common/Exception.h:109: DB::Exception::createRuntime(int, String&) @ 0x00000000064a042c
16:32:19  2. ../local-engine/jni/jni_common.h:98: unsigned char local_engine::safeCallBooleanMethod<>(JNIEnv_*, _jobject*, _jmethodID*) @ 0x00000000064a0f5f
16:32:19  3. /home/jenkins/agent/workspace/gluten/gluten-ci/gluten/cpp-ch/build/../local-engine/Shuffle/NativeSplitter.cpp:161: local_engine::NativeSplitter::hasNext() @ 0x000000000c12c1e5
16:32:19  4. /home/jenkins/agent/workspace/gluten/gluten-ci/gluten/cpp-ch/build/../local-engine/local_engine_jni.cpp:1147: Java_org_apache_gluten_vectorized_BlockSplitIterator_nativeHasNext @ 0x000000000649ab37
16:32:19  
16:32:19  	at org.apache.gluten.vectorized.BlockSplitIterator.nativeHasNext(Native Method)
16:32:19  	at org.apache.gluten.vectorized.BlockSplitIterator.hasNext(BlockSplitIterator.java:55)
16:32:19  	at org.apache.spark.sql.execution.utils.CHExecUtil$$anon$2.hasNext(CHExecUtil.scala:165)
16:32:19  	at org.apache.gluten.vectorized.CloseablePartitionedBlockIterator.hasNext(CloseablePartitionedBlockIterator.scala:33)
16:32:19  	at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:140)
16:32:19  	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
16:32:19  	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
16:32:19  	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
16:32:19  	at org.apache.spark.scheduler.Task.run(Task.scala:136)
16:32:19  	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:548)
16:32:19  	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1504)
16:32:19  	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:551)
16:32:19  	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
16:32:19  	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
16:32:19  	at java.lang.Thread.run(Thread.java:748)
```

**Comment by wForget** [comment]
> Could you please help to see if the [failed GA](https://opencicd.kyligence.com/job/gluten/job/gluten-ci/11653/) is a relevant failure?

It doesn't seem to be a related failure. I retriggered GA and the previously failed task passed.

**Comment by wForget** [comment]
@zhztheplayer @PHILO-HE Could you please continue looking at it?

**Review by zhztheplayer — COMMENTED** [review]
Thanks! Overall looking good.

BTW, is it possible to have a unit test to check on Gluten's behavior when os platform/arch is not supported by Jar? Perhaps we can raise an error like `... not supported by current Gluten build` in that case?

**Review by zhztheplayer — APPROVED** [review]

File: gluten-substrait/src/main/java/org/apache/gluten/vectorized/JniLibLoader.java [hunk 3]
```diff
@@ -158,11 +158,10 @@ public static void unloadFromPath(String libPath) {
     }
   }
 
-  public void mapAndLoad(String unmappedLibName, boolean requireUnload) {
+  public void mapAndLoad(String unmappedLibPath, boolean requireUnload) {
```
[the remark this answers is not in the document]
Comment (wForget) [reply gluten-substrait/src/main/java/org/apache/gluten/vectorized/JniLibLoader.java:161]: > Can we remove the method now?

It seems possible, I will remove it.

**Comment by wForget** [comment]
> BTW, is it possible to have a unit test to check on Gluten's behavior when os platform/arch is not supported by Jar? Perhaps we can raise an error like ... not supported by current Gluten build in that case?

I will try to add a checker in `buildbundle-veloxbe.sh` to check if it is a supported platform and arch.

**Review by zhztheplayer — APPROVED** [review]

**Review by PHILO-HE — APPROVED** [review]

**Pull request merged**